src/tools/snapshot/nodes/ObjectSerializationNodes.java

```diff
@@ -0,0 +1,274 @@
+                DispatchGuard.createSObjectCheck(o),
+                next,
+                false);
+      }
+
+      return reads;
+    }
+  }
+
+  protected static final class UninitializedObjectSerializationNode
+      extends ObjectSerializationNode {
+
+    private UninitializedObjectSerializationNode(final SClass clazz) {
+      super(clazz);
+    }
+
+    @Override
+    public void serialize(final Object o, final SnapshotBuffer sb) {
+      if (o instanceof SObject) {
+        clazz.setSerializer(
+            SObjectSerializationNode.create(clazz, createReadNodes((SObject) o)));
+        clazz.getSerializer().serialize(o, sb);
+      } else if (o instanceof SObjectWithoutFields) {
+        clazz.setSerializer(
+            new SObjectWithoutFieldsSerializationNode(clazz));
+        clazz.getSerializer().serialize(o, sb);
+      }
+    }
+
+    @Override
+    public Object deserialize(final ByteBuffer sb) {
+      if (clazz.getFactory().hasSlots()) {
+        clazz.setSerializer(
+            SObjectSerializationNode.create(clazz));
+        return clazz.getSerializer().deserialize(sb);
+      } else {
+        clazz.setSerializer(
+            new SObjectWithoutFieldsSerializationNode(clazz));
+        return clazz.getSerializer().deserialize(sb);
+      }
+    }
+  }
+
+  // These Nodes may need to be replaced in the class when object layouts change
+
+  @GenerateNodeFactory
+  public abstract static class SObjectSerializationNode
+      extends ObjectSerializationNode {
+
+    protected final int                         fieldCnt;
+    @Children private CachedSlotRead[]          fieldReads;
+    @Children private CachedSlotWrite[]         fieldWrites;
+    @Children private CachedSerializationNode[] cachedSerializers;
+    protected ObjectLayout                      layout;
+
+    protected abstract void execute(SObject o, SnapshotBuffer sb);
+
+    protected SObjectSerializationNode(final SClass clazz,
+        final CachedSlotRead[] reads) {
+      super(clazz);
+      layout = clazz.getLayoutForInstancesUnsafe();
+      fieldReads = insert(reads);
+      fieldCnt = fieldReads.length;
+    }
+
+    public static SObjectSerializationNode create(final SClass clazz,
+        final CachedSlotRead[] reads) {
+      return SObjectSerializationNodeFactory.create(clazz, reads);
+    }
+
+    @Override
+    public void serialize(final Object o, final SnapshotBuffer sb) {
+      assert o instanceof SObject;
+
+      SObject so = (SObject) o;
+      if (!so.isLayoutCurrent()) {
+        ObjectTransitionSafepoint.INSTANCE.transitionObject(so);
+      }
+
+      if (!layout.isValid()) {
+        // deoptimize and adopt read nodes that fit the current layout
+        CompilerDirectives.transferToInterpreterAndInvalidate();
+        layout = clazz.getLayoutForInstancesUnsafe();
+        CachedSlotRead[] newChildren = createReadNodes(so);
+
+        for (int i = 0; i < fieldCnt; i++) {
+          fieldReads[i].replace(newChildren[i]);
```

Comment:
this looks brittle and I can't tell whether this even works.

I was wondering whether simply replacing the full node (`this`) would be saver?